ext/spl/spl_fixedarray.c

```diff
@@ -754,6 +748,120 @@ PHP_METHOD(SplFixedArray, getIterator)
 	zend_create_internal_iterator_zval(return_value, ZEND_THIS);
 }
 
+/* Assumes the object has been created and the spl_fixedarray_object's
+ * array member is uninitialized or zero-initialized.
+ * The state can have string keys, but they must come prior to integers:
+ *
+ * 	SplFixedArray::__set_state(array(
+ * 	   'property' => 'value',
+ * 	   0 => 1,
+ * 	   1 => 2,
+ * 	   2 => 3,
+ * 	))
+ */
+static zend_result spl_fixedarray_import(zend_object *object, HashTable *ht)
+{
+	spl_fixedarray_object *intern = spl_fixed_array_from_obj(object);
+	spl_fixedarray *fixed = &intern->array;
+
+	spl_fixedarray_init(fixed, 0);
+
+	/* For performance we do a single pass over the input array and the
+	 * spl_fixedarray elems. This complicates the implementation, but part
+	 * of the reason for choosing SplFixedArray is for peformance, and
+	 * although that is primarily for memory we should still care about CPU.
+	 *
+	 * We need to track the number of string keys, which must come before
+	 * all the integer keys. This way the total size of the input minus the
+	 * number of string keys equals the number of integer keys, so we can
+	 * allocate the spl_fixedarray.elements and do this in a single pass.
+	 */
+
+	// The total size of the input array
+	const zend_long nht = zend_hash_num_elements(ht);
+	zend_long nstrings = 0; // The number of string keys
+	zend_long nints; // will be nht - nstrings
+
+	zend_string *str_idx = NULL; // current string key of input array
+	zend_ulong num_idx = 0; // current int key (valid iff str_idx == NULL)
+	zval *val = NULL; // current value of input array
+	zend_long max_idx = -1; // the largest index found so far
+	zval *begin = NULL; // pointer to beginning of fixedarray's storage
+	zval *end = NULL; // points one element passed the last element
+	const char *ex_msg = NULL; // message for the value exception
+
+	ZEND_HASH_FOREACH_KEY_VAL(ht, num_idx, str_idx, val) {
```

Comment:
```suggestion
	ZEND_HASH_FOREACH_KEY_VAL(ht, num_idx, str_idx, val) {
			ZVAL_DEREF(val);
```

This should either dereference references or throw an exception, e.g. for `$arr = [&$val]; SplFixedArray::__set_state($arr);` there shouldn't be references in the final result.